memlimit/memlimit.go

```diff
@@ -29,56 +30,106 @@ var (
 	logger = log.New(io.Discard, "", log.LstdFlags)
 )
 
+type config struct {
+	printf   func(string, ...interface{})
+	ratio    float64
+	provider Provider
+}
+
+// An Option alters the behavior of SetGoMemLimitWithEnv.
+type Option interface {
+	apply(*config)
+}
+
+type optionFunc func(*config)
+
+func (of optionFunc) apply(cfg *config) { of(cfg) }
+
+// WithLogger uses the supplied printf implementation for log output.
+// By default log.Printf.
+func WithLogger(printf func(string, ...interface{})) Option {
+	return optionFunc(func(cfg *config) {
+		cfg.printf = printf
+	})
+}
+
+// WithRatio configure memory limit.
+// By default `defaultAUTOMEMLIMIT`.
+func WithRatio(ratio float64) Option {
+	return optionFunc(func(cfg *config) {
+		cfg.ratio = ratio
+	})
+}
+
+// WithProvider configure provider.
+// By default `FromCgroup`.
+func WithProvider(provider Provider) Option {
+	return optionFunc(func(cfg *config) {
+		cfg.provider = provider
+	})
+}
+
 // SetGoMemLimitWithEnv sets GOMEMLIMIT with the value from the environment variable.
 // You can configure how much memory of the cgroup's memory limit to set as GOMEMLIMIT
 // through AUTOMEMLIMIT in the half-open range (0.0,1.0].
 //
 // If AUTOMEMLIMIT is not set, it defaults to 0.9. (10% is the headroom for memory sources the Go runtime is unaware of.)
 // If GOMEMLIMIT is already set or AUTOMEMLIMIT=off, this function does nothing.
-func SetGoMemLimitWithEnv() {
+func SetGoMemLimitWithEnv(opts ...Option) {
```

Comment:
Changing the signature of `SetGoMemLimitWithEnv` makes a breaking change, and I'd like to avoid it as much as possible, though the current version is under `v1.0.0`.

What about adding a new function like the one below and merging them in the `v1.0.0` release?

```suggestion
func SetGoMemLimitWithEnv() {
	SetGoMemLimitWithOpts()
}

func SetGoMemLimitWithOpts(opts ...Option) {
```

Reply:
@KimMachineGun, I propose a new feature `SetGoMemLimitWithOptions`